### ISSUE TITLE
Reduced performance in long line drawing in v2.6.0

### ISSUE THREAD

**Issue by yaugenka** [openstreetmap/iD#4731]
There are notable lags when editing long ways (1k+ nodes) in v2.6.0.
Prior to the update the reduced berformance was observed only when you have that much way nodes unsaved. After the update it is lagging even when you are editing a just saved line.

![lags](https://user-images.githubusercontent.com/14937001/35247566-faa7b466-ffdb-11e7-83ee-22b301fc86fb.gif)

**Comment by bhousel** [comment]
Thanks for reporting.  I'll see if we can optimize the line self-intersection checks. 
Another idea - it might be enough to just check the most recent 100 points or something.

**Comment by yaugenka** [comment]
@bhousel is it really that important to do the self-intersection checks at runtime? won't it be enough to do it at save?

**Comment by bhousel** [comment]
> is it really that important to do the self-intersection checks at runtime? won't it be enough to do it at save?

Yes, I think it is important for the user to get immediate feedback that how they are drawing a line is wrong..  Could you imagine if you edited for an hour and then tried to save and found out half your lines were wrong?  That would make me so sad  😢

**Comment by yaugenka** [comment]
If iD could list and let you navigate to the intersection nodes similar to how untagged lines are handled it wouldn't be a big deal to fix the mistakes to my mind.

**Comment by slhh** [comment]
@bhousel The current code seems to check the full line for any intersection, which is quite expensive.
For placing a new node you do only need to check for intersections of the new edge with the already draw line.

**Comment by bhousel** [comment]
> The current code seems to check the full line for any intersection, which is quite expensive.
> For placing a new node you do only need to check for intersections of the new edge with the already draw line.

That's true, but I don't think the code knows the drawn segments - D3 clips the line for us.

I do think there are probably opportunities to speed this up.  Snapping a point to a line is also a similar algorithm, (it projects the point onto every segment), and that one works fast.
